### PR TITLE
Feedback Sections on SO - SEB

### PR DESCRIPTION
**Pull request by seb-odoo**
Task https://www.odoo.com/web#id=1869469&action=333&active_id=1217&model=project.task&view_type=form&menu_id=4720

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Hi, Sébastien, one of the things I missed from one2many "add" buttons right now is the possibility of having shortcut keys for pressing directly to add them, so the keyboard access enhancement gets reduced in this case. Navigating with tab in Sales Orders, you get the focus on "Add a note" button, with no possibility of going to "Add a product" without the mouse. Have you considered this?

**Comment by seb-odoo** [comment]
@pedrobaeza thanks for the feedback, we will fix the tab navigation issue as it obviously should focus add a product and add a section too.

Did add a line previously also had a shortcut key?

**Comment by pedrobaeza** [comment]
I didn't see it before your previous improvement, but I think it should be there, don't you think? (but maybe is not your duty then...)

File: addons/account/views/account_portal_templates.xml [hunk 1]
```diff
@@ -77,16 +81,16 @@
...
                         </div>
                         <div class="o_payment_button">
                             <div class="text-center mb8 ml16 mr16 o_download_pdf">
-                                <a class="btn btn-secondary btn-block o_download_btn" t-att-href="'/my/invoices/pdf/%s?%s' % (invoice.id, keep_query())" title="Download"><i class="fa fa-download"/> Download</a>
-                                <a class="btn btn-secondary btn-block d-none d-lg-inline-block o_print_btn o_portal_invoice_print" t-att-href="'/my/invoices/pdf/%s?%s&amp;print=true' % (invoice.id, keep_query())" id="print_invoice_report" title="Print"><i class="fa fa-print"/> Print</a>
+                                <a class="btn btn-secondary btn-block o_download_btn" t-att-href="invoice.get_portal_url() + '&amp;report_type=pdf'" title="Download"><i class="fa fa-download"/> Download</a>
+                                <a class="btn btn-secondary btn-block d-none d-lg-inline-block o_print_btn o_portal_invoice_print" t-att-href="invoice.get_portal_url() + '&amp;report_type=pdf&amp;print=true'" id="print_invoice_report" title="Print"><i class="fa fa-print"/> Print</a>
```
Comment (seb-odoo) [inline addons/account/views/account_portal_templates.xml]: replace print by download,

File: addons/portal/controllers/portal.py [hunk 2]
```diff
@@ -264,3 +266,23 @@ def _get_page_view_values(self, document, access_token, values, session_history,
         values.update(get_records_pager(history, document))
 
         return values
+
+    def _show_report(self, model, report_type, report_ref, **kw):
+        if report_type in ('html', 'pdf', 'text'):
+            # print report as sudo, since it require access to taxes, payment term, ... and portal
+            # does not have those access rights.
+            report_sudo = request.env.ref(report_ref).sudo()
```
Comment (seb-odoo) [inline addons/portal/controllers/portal.py]: check this is of type report

File: addons/portal/controllers/portal.py [hunk 3]
```diff
@@ -264,3 +266,23 @@ def _get_page_view_values(self, document, access_token, values, session_history,
...
+
+            method_name = 'render_qweb_%s' % (report_type)
+            if hasattr(report_sudo, method_name):
```
Comment (seb-odoo) [inline addons/portal/controllers/portal.py]: remove the check

File: addons/portal/controllers/portal.py [hunk 4]
```diff
@@ -264,3 +266,23 @@ def _get_page_view_values(self, document, access_token, values, session_history,
...
+                invoice_report = getattr(report_sudo, method_name)([model.id], data={'report_type': report_type})[0]
+                reporthttpheaders = [
+                    ('Content-Type', 'application/pdf' if report_type == 'pdf' else 'text/html'),
+                    ('Content-Length', len(invoice_report)),
+                ]
+                if report_type == 'pdf' and not kw.get('print'):
+                    filename = "%s.pdf" % (re.sub('\W+', '-', model._get_report_filename()))
+                    reporthttpheaders.append(('Content-Disposition', content_disposition(filename)))
+                return request.make_response(invoice_report, headers=reporthttpheaders)
+            raise UserError("Model %s has no method: %s" % (report_sudo, method_name))
```
Comment (seb-odoo) [inline addons/portal/controllers/portal.py]: remove

File: addons/portal/controllers/portal.py [hunk 5]
```diff
@@ -264,3 +266,23 @@ def _get_page_view_values(self, document, access_token, values, session_history,
...
+            report_sudo = request.env.ref(report_ref).sudo()
+
+            method_name = 'render_qweb_%s' % (report_type)
+            if hasattr(report_sudo, method_name):
+                invoice_report = getattr(report_sudo, method_name)([model.id], data={'report_type': report_type})[0]
+                reporthttpheaders = [
+                    ('Content-Type', 'application/pdf' if report_type == 'pdf' else 'text/html'),
+                    ('Content-Length', len(invoice_report)),
+                ]
+                if report_type == 'pdf' and not kw.get('print'):
```
Comment (seb-odoo) [inline addons/portal/controllers/portal.py]: download instead of print

File: addons/sale/controllers/portal.py [hunk 6]
```diff
@@ -162,18 +160,17 @@ def portal_order_page(self, order_id, pdf=None, access_token=None, message=False
...
             'message': int(message) if message else False,
-            'action': request.env.ref('sale.action_quotations').id,
             'no_breadcrumbs': request.env.user.partner_id.commercial_partner_id not in order_sudo.message_partner_ids,
             'tx_state': transaction.state if transaction else False,
             'need_payment': order_sudo.invoice_status == 'to invoice' and transaction.state in ['draft', 'cancel'],
             'token': access_token,
             'return_url': '/shop/payment/validate',
             'bootstrap_formatting': True,
             'partner_id': order_sudo.partner_id.id,
+            'type_html': True,
```
Comment (seb-odoo) [inline addons/sale/controllers/portal.py]: use report_type? if in this case here, report type is probably undefined

File: addons/sale/models/sale.py [hunk 7]
```diff
@@ -833,6 +841,12 @@ def get_portal_last_transaction(self):
     def _get_customer_lead(self, product_tmpl_id):
         return False
 
+    @api.multi
+    def _get_report_filename(self):
```
Comment (seb-odoo) [inline addons/sale/models/sale.py]: rename to _get_report_base_filename

File: addons/sale/views/sale_portal_templates.xml [hunk 8]
```diff
@@ -376,17 +387,15 @@
...
+                <span t-field="sale_order.type_name"></span>
                 <em t-esc="sale_order.name"/>
                 <div t-ignore="true" class="float-md-right css_editable_mode_hidden d-print-none">
                     <ul class="list-inline">
                         <li groups="sales_team.group_sale_salesman" class="list-inline-item">
-                            <a role="button" t-if="sale_order.state not in ('manual')" class="btn btn-info" t-att-href="'/web#return_label=Website&amp;model=%s&amp;id=%s&amp;action=%s&amp;view_type=form' % (sale_order._name, sale_order.id, action)">Edit Quotation</a>
-                            <a role="button" t-if="sale_order.state in ('manual')" class=" btn btn-info" t-att-href="'/web#return_label=Website&amp;model=%s&amp;id=%s&amp;action=%s&amp;view_type=form' % (sale_order._name, sale_order.id, action)">Back to Sales Order</a>
+                            <a role="button" class="btn btn-info" t-att-href="'/web#return_label=Website&amp;model=%s&amp;id=%s&amp;action=%s&amp;view_type=form' % (sale_order._name, sale_order.id, sale_order.env.ref('sale.action_quotations').id)">Edit <span t-field="sale_order.type_name"></span></a>
```
Comment (seb-odoo) [inline addons/sale/views/sale_portal_templates.xml]: translation: keep the whole text "edit **type**"

File: addons/sale/views/sale_portal_templates.xml [hunk 9]
```diff
@@ -376,17 +387,15 @@
...
                         </li>
                         <li class="list-inline-item">
-                            <a role="button" class="btn btn-info" target="_blank" t-att-href="sale_order.get_portal_url() + '&amp;pdf=True'">Print</a>
+                            <a role="button" class="btn btn-info" target="_blank" t-att-href="sale_order.get_portal_url() + '&amp;report_type=pdf&amp;print=true'">Print</a>
```
Comment (seb-odoo) [inline addons/sale/views/sale_portal_templates.xml]: use named params

**Pull request merged**